src/main/java/seedu/address/commons/util/CsvFile.java

```diff
@@ -0,0 +1,197 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import seedu.address.commons.exceptions.CsvMismatchedColumnException;
+import seedu.address.commons.exceptions.CsvMissingFieldException;
+
+/**
+ * Representation of the contents of a CSV file as an object.
+ */
+public class CsvFile {
+    private final int numColumns;
+
+    private final String delimiter;
+    private final HashMap<String, Integer> columnIndices;
+    private final String header;
+    private final List<CsvRow> rows;
+
+    /**
+     * Represents a row of values in the CSV file. Encapsulating the row values
+     * within this object allows one to extract the value in a column with just
+     * the column's name, without having to know the column index.
+     */
+    public class CsvRow implements GetValuer {
+        private final List<String> vals;
+
+        /**
+         * Constructs a CsvRow object. If the number of values supplied in the row is less
+         * than the number of columns in the file, the values will be regarded as the values for
+         * the first n columns, where n is the number of values in the row. The remaining columns
+         * will be treated as empty values (empty string).
+         * @param values String containing the values of the row delimited by the CSV delimiter
+         * @throws CsvMismatchedColumnException Thrown if the number of values in the row is greater
+         *      than the number of columns in the file
+         */
+        public CsvRow(String values) throws CsvMismatchedColumnException {
+            String[] row = values.split(delimiter);
+            boolean hasTooManyColumns = row.length > numColumns;
+            if (hasTooManyColumns) {
+                throw new CsvMismatchedColumnException(numColumns, row.length);
+            }
+
+            vals = new ArrayList<>(Arrays.asList(row));
+            padRow();
+        }
+
+        /**
+         * Constructs a CsvRow object. If the number of values supplied in the row is less
+         * than the number of columns in the file, the values will be regarded as the values for
+         * the first n columns, where n is the number of values in the row. The remaining columns
+         * will be treated as empty values (empty string).
+         * @param values List containing the values of the row delimited by the CSV delimiter
+         * @throws CsvMismatchedColumnException if the number of values in the row is greater
+         *      than the number of columns in the file
+         */
+        public CsvRow(List<?> values) throws CsvMismatchedColumnException {
+            boolean hasTooManyColumns = values.size() > numColumns;
+            if (hasTooManyColumns) {
+                throw new CsvMismatchedColumnException(numColumns, values.size());
+            }
+
+            vals = values.stream().map(Object::toString).collect(Collectors.toList());
+            padRow();
+        }
```

Comment:
Just a quick clarification, because i do understand the CsvRow constructor that parse in a string, because that is generally how the csv file is processed. I am slightly more confused when it comes to this constructor here. Is this to account for the reverse? In the sense that we have an entry of Person for example, and we want to convert it to a CsvRow object? If so then i wish to clarify about the use of List<?>, because not every list of objects can be converted in to a row in a csv file. Same with Line 172.

Should we limit this to only take in Persons instead? Could we create an interface called CsvParsable

Reply:
CsvFile is intended to be an intermediate stage between objects and the raw CSV file. This constructor targets the object -> CSV process. Since the delimiter can be different for different files e.g. we're using semicolons in this instance, but it could very well be colons/pipes/etc., I didn't want the object to provide its string serialisation as that would require knowledge of the delimiter. As such, I chose to take in a list of values as input so that the CsvRow can handle the serialisation by referring to the delimiter specified in the parent CsvFile.

Implementing CsvParsable seems like an interesting proposal, and I'll look into modifying it in the next iteration. However, the parsable method will still have to return a list of values for the abovementioned reason.

Reply:
Awesome, thank you for the clarification.